Neo4j.Driver/Neo4j.Driver/Internal/Session.cs

```diff
@@ -68,97 +98,104 @@ protected override void Dispose(bool isDisposing)
+        /// </summary>
+        private void RunResultCleanupCallback()
         {
-            return TryExecute(() =>
-            {
-                lock (_txSyncLock)
-                {
-                    EnsureCanRunMoreStatements();
-                    _transaction = new Transaction(_connection, _transactionCleanupAction, _logger, bookmark);
-                }
-                return _transaction;
-            });
+            Throw.ArgumentNullException.IfNull(_sessionRunResult, nameof(_sessionRunResult));
+            Throw.ArgumentNullException.IfNull(_connection, nameof(_connection));
+
+            CleanRunResultResources();
         }
 
-        public string LastBookmark { get; private set; }
+        private void CleanRunResultResources()
+        {
+            LastBookmark = null;
+            _sessionRunResult = null;
 
-        private void EnsureCanRunMoreStatements()
+            // always try to close connection used by the result too
+            _connection?.Dispose();
+            _connection = null;
+        }
+
+        /// <summary>
+        /// Called back in <see cref="Transaction.Dispose"/>
+        /// </summary>
+        private void TransactionCleanupCallback()
         {
-            EnsureConnectionIsHealthy();
-            EnsureNoOpenTransaction();
-            EnsureSessionIsOpen();
+            Throw.ArgumentNullException.IfNull(_transaction, nameof(_transaction));
+            Throw.ArgumentNullException.IfNull(_connection, nameof(_connection));
+
+            LastBookmark = _transaction.Bookmark;
+            _transaction = null;
+
+            // always dispose connection used by the transaction too
+            _connection.Dispose();
+            _connection = null;
         }
 
-        private void EnsureSessionIsOpen()
+        /// <summary>
+        /// Clean any session.run result reference.
+        /// If session.run result is not fully consumed, then pull full result into memory.
+        /// </summary>
+        /// <exception cref="ClientException">If error when pulling result into memory</exception>
+        private void DisposeSessionResult()
         {
-            if (!_isOpen)
+            if (_sessionRunResult != null)
             {
-                throw new ClientException("Cannot running more statements in the current session as it has already been disposed." +
-                                          "Make sure that you do not have a bad reference to a disposed session " +
-                                          "and retry your statement in another new session.");
+                try
+                {
+                    // this will enfore to buffer all unconsumed result
+                    _connection.Sync();
+                }
+                catch (Exception e)
+                {
+                    throw new ClientException($"Error when pulling unconsumed session.run records into memory in session: {e.Message}", e);
```

Comment:
Would it be strange that you get the error of previous run in a new run? This wrapper is for not confusing a user where the error comes from and also try to educate the user that he should consume his result.